### PR TITLE
fix: Fetch local fingerprint from SDP

### PR DESCRIPTION
**Pull request by ckousik**
Since firefox does not support `getFingerprints`, we fetch the local fingerprint from the PeerConnection's `localDescription` field by parsing the fingerprint line from the SDP.

### CODE REVIEW

File: src/sdp.ts
Comment (mxinden) [inline src/sdp.ts:13]: Left over to be removed, right?

File: src/transport.ts
Comment (mxinden) [inline src/transport.ts:217]: Not sure how this is handled in the js-libp2p codebase in general. Should we still use `localCert.getFingerprints` in case the API is available (e.g. on Chrome), or should we always use the hack through the SDP?

**Review by mxinden — COMMENTED** [review]
Thanks for tackling this.

**Comment by mxinden** [comment]
@achingbrain @MarcoPolo any objections to merging here?

File: src/transport.ts
Comment (MarcoPolo) [reply src/transport.ts:217]: I think I would prefer trying to use the best option and falling back on the hack. But not blocking this.

**Review by MarcoPolo — COMMENTED** [review]
Can we have firefox run this test in CI?

File: src/transport.ts
Comment (SgtPooki) [reply src/transport.ts:218]: I agree with Marco here, but it would be good to know (debug logs) which option was used if we're going to have different paths.

File: src/sdp.ts
Comment (SgtPooki) [inline src/sdp.ts:34]: Prefer `regex.match`.

Using `exec` can cause unexpected issues with the stored `lastIndex` being updated and then causing match failures in later `.exec` calls. 

> JavaScript [RegExp](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp) objects are stateful when they have the [global](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/global) or [sticky](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) flags set (e.g. /foo/g or /foo/y). They store a [lastIndex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/lastIndex) from the previous match. Using this internally, exec() can be used to iterate over multiple matches in a string of text (with capture groups), as opposed to getting just the matching strings with [String.prototype.match()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match).

see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description

Also see https://stackoverflow.com/a/3811949/592760

**Review by SgtPooki — CHANGES_REQUESTED** [review]

File: src/sdp.ts
Comment (SgtPooki) [inline src/sdp.ts:28]:
```suggestion
const fingerprintRegex = /^a=fingerprint:(?:\w+-[0-9]+)\s(?<fingerprint>(:?[0-9a-fA-F]{2})+)$/m
function getFingerprintFromSdp (sdp: string): string | undefined {
  const searchResult = sdp.match(fingerprintRegex)
```

File: src/sdp.ts
Comment (ckousik) [reply src/sdp.ts]: Thanks @SgtPooki . I was not aware of this. I've incorporated the change and added a simple test.

File: src/sdp.ts
Comment (ckousik) [reply src/sdp.ts]: Fixed

File: src/transport.ts
Comment (ckousik) [reply src/transport.ts:218]: @SgtPooki I've added the fallback to fetching from the SDP as requested.

**Comment by p-shahi** [comment]
@MarcoPolo ping to re-review

**Comment by ckousik** [comment]
@MarcoPolo I'm still working on firefox browser-to-server tests in CI. Could we move that work to a subsequent PR?

**Comment by ckousik** [comment]
@MarcoPolo The browser-to-server tests is running on Firefox in CI.

File: src/sdp.ts
Comment (SgtPooki) [inline src/sdp.ts:52]: `'' == null // false`

Returning `''` on line 49 will cause the check on line 215 in src/transports.ts, `if (localFingerprint == null) {` to return false. We should return undefined here, or even better:

```suggestion
  const searchResult = sdp.match(fingerprintRegex)

  return searchResult?.groups?.fingerprint
```

File: test/sdp.spec.ts
Comment (SgtPooki) [inline test/sdp.spec.ts:56]: we should add a negative test here to validate the fallback behaviour

**Review by SgtPooki — CHANGES_REQUESTED** [review]
i'm not an expert in the business logic of what is going on here, but i ran through the code and found a single potential failure that should be addressed.

File: src/sdp.ts
Comment (ckousik) [reply src/sdp.ts:52]: Thank you!

File: test/sdp.spec.ts
Comment (ckousik) [reply test/sdp.spec.ts:56]: this just tests the regex behaviour. While there are no explicit tests for the fallback behaviour, it gets tested when the tests are run in Firefox as part of CI.

**Review by SgtPooki — COMMENTED** [review]
Lgtm but i’ll defer to primary maintainers for final approval

**Comment by p-shahi** [comment]
@MarcoPolo bump!

File: src/transport.ts
Comment (MarcoPolo) [inline src/transport.ts]: This may copy unnecessarily (I know this was here before).

**Review by MarcoPolo — APPROVED** [review]

File: examples/browser-to-server/package.json
Comment (MarcoPolo) [reply examples/browser-to-server/package.json:12]: `aegir test` usually builds before testing. This is in line with that. We should probably build in the `test:chrome` as well.

Would be nice to have a test script that builds once then runs all the tests.

File: src/stream.ts
Comment (MarcoPolo) [reply src/stream.ts:216]: It should be since the default is supposedly a blob: https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/binaryType. However I see the binaryType as arraybuffer when I create the channel. So I'm not sure why. I think we should keep this anyways.

**Comment by MarcoPolo** [comment]
Thanks a lot @SgtPooki! Your review was very helpful :)

And thanks @wemeetagain for taking a look

**Review by MarcoPolo — APPROVED** [review]

**Pull request merged**
